sotodlib/tod_ops/detrend.py

```diff
@@ -52,14 +53,24 @@ def detrend_data(tod, method='linear', axis_name='samps',
         x = np.linspace(0, 1, n_samps)
         count = max(1, min(count, signal.shape[-1] // 2))
         slopes = signal[...,-count:].mean(axis=-1)-signal[...,:count].mean(axis=-1)
-        signal = signal - slopes[...,None] * x
+        ## the 2d loop is significantly faster if possible
+        if len(signal.shape) == 2:
+            for i in range(signal.shape[0]):
+                signal[i,:] -= slopes[i]*x
+        else:
+            signal -= slopes[...,None] * x
         signal -= np.mean(signal, axis=-1)[...,None]
     else:
         raise ValueError("method flag must be linear, mean, or median")
 
     if axis_idx != signal.ndim - 1:
         signal = signal.transpose(tuple(axis_reorder))
-    return signal.astype(dtype_in)
+
+    if signal.dtype != dtype_in:
+        # .astype takes awhile on long arrays, only cast if necessary
+        return signal.astype(dtype_in)
+    else:
+        return signal
```

Comment:
It might be worth guarding against the first case entirely.  If you construct x and slopes to have the same type as signal, that should be enough.  Then you can assert signal.dtype == dtype_in as a fail-safe that will hopefully never ring.